lib/controllers/messages.ts

```diff
@@ -233,33 +233,64 @@ async function getMessageNotificationStatuses(
           retrievedMessage.id
         }`
       );
-      return right(Promise.resolve({}));
+      return right({});
     }
     const notification = maybeNotification.value;
     // Get NotificationStatus looping on all notification channels
-    return right(
-      Object.keys(NotificationChannelEnum).reduce(
-        async (
-          statusesP: Promise<NotificationStatusHolder>,
-          channelIdx: string
-        ): Promise<NotificationStatusHolder> => {
-          const channel =
-            NotificationChannelEnum[channelIdx as NotificationChannelEnum];
-          const errorOrMaybeStatus = await notificationStatusModel.findOneNotificationStatusByNotificationChannel(
-            notification.id,
-            channel
-          );
-          const statusObj = fromEither(errorOrMaybeStatus)
-            .chain(t.identity)
-            .toUndefined();
-          const statuses = await statusesP;
-          return statusObj
-            ? { ...statuses, [channel.toLowerCase()]: statusObj.status }
-            : statuses;
-        },
-        Promise.resolve({})
-      )
+    /*
+    const channelStatuses = channelKeys.reduce(
+      async (
+        statusesP: Promise<NotificationStatusHolder>,
+        channelIdx: string
+      ): Promise<NotificationStatusHolder> => {
+        const channel =
+          NotificationChannelEnum[channelIdx as NotificationChannelEnum];
+        const errorOrMaybeStatus = await notificationStatusModel.findOneNotificationStatusByNotificationChannel(
+          notification.id,
+          channel
+        );
+        const statusObj = fromEither(errorOrMaybeStatus)
+          .chain(t.identity)
+          .toUndefined();
+        const statuses = await statusesP;
+        return statusObj
+          ? { ...statuses, [channel.toLowerCase()]: statusObj.status }
+          : statuses;
+      },
+      Promise.resolve({})
+    );
+    */
+
+    // returns the status of a channel
+    const getChannelStatus = async (channel: NotificationChannelEnum) => {
```

Comment:
ahh nice, I wasn't super enthusiast of the code indeed : )

Reply:
I would also try to avoid using `Object.keys` + `Array#reduce` and explicitly enumerate the keys (channels) - see also my comment on `pick` - the mindset you should have is that the information about object structure (interfaces) is not available at runtime, so any structural inspection/change that happens at runtime should be avoided (it's kind of like java erasure/introspection)

Reply:
mmm, that defeats the whole purpose of the refactor which is abstract away how many channels are implemented, so you don't need to have to find / add a new channel everywhere in the code (models, controllers, etc.) everytime you add one. it's more like polymorphism than introspection. imho, as far as the array objects implements the same interface (ie. they are strings from enums) it's ok to cycle on them with keys / reduce. you can add typings in the reducer declaration (which is what we usually do) to be more strict.